benchmarks/babelstream/src/babelStreamMainTest.cpp

```diff
@@ -267,210 +296,328 @@ void testKernels()
-            auto start = std::chrono::high_resolution_clock::now();
-            kernelFunc();
-            alpaka::wait(queue);
-            auto end = std::chrono::high_resolution_clock::now();
-            std::chrono::duration<double> duration = end - start;
-            runtime = duration.count();
-            times.push_back(runtime);
+            runtimeResults.addKernelTimingsVec("DotKernel");
         }
+    }
+    else if(kernelsToBeExecuted == KernelsToRun::NStream)
+    {
+        runtimeResults.addKernelTimingsVec("NStreamKernel");
+    }
+    else if(kernelsToBeExecuted == KernelsToRun::Triad)
+    {
+        runtimeResults.addKernelTimingsVec("TriadKernel");
+    }
 
-        // find the minimum of the durations array.
-        // In benchmarking the first item of the runtimes array is not included in calculations.
-        const auto minmaxPair = findMinMax(times);
-        minExecTimesOfKernels.push_back(minmaxPair.first);
-        maxExecTimesOfKernels.push_back(minmaxPair.second);
-        avgExecTimesOfKernels.push_back(findAverage(times));
-        kernelLabels.push_back(kernelLabel);
-        times.clear();
-    };
 
-    // Run kernels one by one
-    // Test the init-kernel.
-    measureKernelExec(
-        [&]()
+    // Main for loop to run the kernel-sequence
+    for(auto i = 0; i < numberOfRuns; i++)
+    {
+        if(kernelsToBeExecuted == KernelsToRun::All)
+        {
+            // Test the copy-kernel. Copy A one by one to C.
+            measureKernelExec(
+                [&]() { alpaka::exec<Acc>(queue, workDivCopy, CopyKernel(), bufAccInputAPtr, bufAccOutputCPtr); },
+                "CopyKernel");
+
+            // Test the scaling-kernel. Calculate B=scalar*C. Where C = A.
+            measureKernelExec(
+                [&]() { alpaka::exec<Acc>(queue, workDivMult, MultKernel(), bufAccOutputCPtr, bufAccInputBPtr); },
+                "MultKernel");
+
+            // Test the addition-kernel. Calculate C=A+B. Where B=scalar*C or B=scalar*A.
+            measureKernelExec(
+                [&]() {
+                    alpaka::exec<Acc>(
+                        queue,
+                        workDivAdd,
+                        AddKernel(),
+                        bufAccInputAPtr,
+                        bufAccInputBPtr,
+                        bufAccOutputCPtr);
+                },
+                "AddKernel");
+        }
+
+        if(kernelsToBeExecuted == KernelsToRun::All || kernelsToBeExecuted == KernelsToRun::Triad)
+        {
+            // Test the Triad-kernel. Calculate A=B+scalar*C. Where C is A+scalar*A.
+            measureKernelExec(
+                [&]() {
+                    alpaka::exec<Acc>(
+                        queue,
+                        workDivTriad,
+                        TriadKernel(),
+                        bufAccInputAPtr,
+                        bufAccInputBPtr,
+                        bufAccOutputCPtr);
+                },
+                "TriadKernel");
```

Comment:
Technically speaking, the upstream version runs [the triad-only version](https://github.com/UoB-HPC/BabelStream/blob/2f00dfb7f8b7cfe8c53d20d5c770bccbf8673440/src/main.cpp#L155) of the code differently. The run-all version measures each run individually, the triad-only version measures the time for the whole loop. Admittedly, that is [fixed in the development branch](https://github.com/UoB-HPC/BabelStream/pull/186) already, so we are practically ahead of our time here.